Git-commit: d0ef798087209d3fdbfd49fbbe5c506f87b4d4cd
From: chenjiajun <chenjiajun8@huawei.com>
Date: Wed, 7 Apr 2021 15:11:46 +0800
Subject: [PATCH] kvm: debugfs: add EXIT_REASON_PREEMPTION_TIMER to vcpu_stat
Patch-mainline: Queued in openEuler repo, version 5.10.0-4.20.0
References: openEuler-21.03
Git-repo: https://gitee.com/openeuler/kernel.git

virt inclusion
category: feature
bugzilla: 46853
CVE: NA

Export EXIT_REASON_PREEMPTION_TIMER kvm exits to vcpu_stat debugfs.
Add a new column to vcpu_stat, and provide preemption_timer status to
virtualization detection tools.

Signed-off-by: chenjiajun <chenjiajun8@huawei.com>
Reviewed-by: Xiangyou Xie <xiexiangyou@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Kai Liu <kai.liu@suse.com>
---
 arch/x86/include/asm/kvm_host.h | 1 +
 arch/x86/kvm/vmx/vmx.c          | 2 ++
 arch/x86/kvm/x86.c              | 1 +
 3 files changed, 4 insertions(+)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index 730ffde044b4..13ba97e63457 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -1070,6 +1070,7 @@ struct kvm_vcpu_stat {
 	u64 utime;
 	u64 stime;
 	u64 gtime;
+	u64 preemption_timer_exits;
 };
 
 struct x86_instruction_info;
diff --git a/arch/x86/kvm/vmx/vmx.c b/arch/x86/kvm/vmx/vmx.c
index 3a0964619070..93a419fe9509 100644
--- a/arch/x86/kvm/vmx/vmx.c
+++ b/arch/x86/kvm/vmx/vmx.c
@@ -5605,6 +5605,7 @@ static fastpath_t handle_fastpath_preemption_timer(struct kvm_vcpu *vcpu)
 
 static int handle_preemption_timer(struct kvm_vcpu *vcpu)
 {
+	++vcpu->stat.preemption_timer_exits;
 	handle_fastpath_preemption_timer(vcpu);
 	return 1;
 }
@@ -6598,6 +6599,7 @@ static fastpath_t vmx_exit_handlers_fastpath(struct kvm_vcpu *vcpu)
 		++vcpu->stat.msr_wr_exits;
 		return handle_fastpath_set_msr_irqoff(vcpu);
 	case EXIT_REASON_PREEMPTION_TIMER:
+		++vcpu->stat.preemption_timer_exits;
 		return handle_fastpath_preemption_timer(vcpu);
 	default:
 		return EXIT_FASTPATH_NONE;
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 2c07fa96ae1f..d387345a41d3 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -285,6 +285,7 @@ struct dfx_kvm_stats_debugfs_item dfx_debugfs_entries[] = {
 	DFX_STAT("utime", utime),
 	DFX_STAT("stime", stime),
 	DFX_STAT("gtime", gtime),
+	DFX_STAT("preemption_timer_exits", preemption_timer_exits),
 	{ NULL }
 };
 
-- 
2.31.1

